gubernator/main.py

```diff
@@ -0,0 +1,239 @@
+        'kubernetes-build',
+        'kubernetes-e2e-gce',
+        'kubernetes-e2e-gce-scalability',
+        'kubernetes-e2e-gce-slow',
+        'kubernetes-e2e-gke',
+        'kubernetes-e2e-gke-slow',
+        'kubernetes-kubemark-5-gce',
+        'kubernetes-kubemark-500-gce',
+        'kubernetes-test-go',
+    }
+}
+
+JINJA_ENVIRONMENT = jinja2.Environment(
+    loader=jinja2.FileSystemLoader(os.path.dirname(__file__) + '/templates'),
+    extensions=['jinja2.ext.autoescape'],
+    trim_blocks=True,
+    autoescape=True)
+JINJA_ENVIRONMENT.line_statement_prefix = '%'
+filters.register(JINJA_ENVIRONMENT.filters)
+
+
+def pad_numbers(s):
+    """Modify a string to make its numbers suitable for natural sorting."""
+    return re.sub(r'\d+', lambda m: m.group(0).rjust(16, '0'), s)
+
+
+def memcache_memoize(prefix, exp=60 * 60, neg_exp=60):
+    """Decorate a function to memoize its results using memcache.
+
+    The function must take a single string as input, and return a pickleable
+    type.
+
+    Args:
+        prefix: A prefix for memcache keys to use for memoization.
+        exp: How long to memoized values, in seconds.
+        neg_exp: How long to memoize falsey values, in seconds
+    Returns:
+        A decorator closure to wrap the function.
+    """
+    # setting the namespace based on the current version prevents different
+    # versions from sharing cache values -- meaning there's no need to worry
+    # about incompatible old key/value pairs
+    namespace = os.environ['CURRENT_VERSION_ID']
+    def wrapper(func):
+        @functools.wraps(func)
+        def wrapped(arg):
+            key = prefix + arg
+            data = memcache.get(key, namespace=namespace)
+            if data is not None:
+                return data
+            else:
+                data = func(arg)
+                if data:
+                    memcache.add(key, data, exp, namespace=namespace)
+                else:
+                    memcache.add(key, data, neg_exp, namespace=namespace)
+                return data
+        return wrapped
+    return wrapper
+
+
+def gcs_read(path):
+    """Read a file from GCS. Returns None on errors."""
+    try:
+        with gcs.open(path) as f:
+            return f.read()
+    except gcs.errors.Error:
+        return None
+
+
+@memcache_memoize('gs-ls://', exp=60)
+def gcs_ls(path):
+    """Enumerate files in a GCS directory. Returns a list of FileStats."""
+    return list(gcs.listbucket(path, delimiter='/'))
+
+
+@memcache_memoize('build-details://', exp=60 * 60 * 4)
+def build_details(build_dir):
+    started = gcs_read(build_dir + '/started.json')
+    finished = gcs_read(build_dir + '/finished.json')
+    if not (started and finished):
+        return
+    started = json.loads(started)
+    finished = json.loads(finished)
+    failures = []
+    for n in xrange(1, 99):
+        junit = gcs_read(
+            '%s/artifacts/junit_%02d.xml' % (build_dir, n))
+        if junit is None:
+            break
```

Comment:
Sometimes we're missing one in the middle if a parallel node doesn't report back, so this logic won't quite work all the time. Not necessary to fix it here, but I'd add a TODO to fix it.